File: frontend/packages/shared/src/domain/entities/Account.ts
import type { BusinessStatusCode } from './Company'

/** 관리자는 서버에서 SQL이나 개발용 로그인으로만 지정됩니다. */
export type AccountRole = 'USER' | 'ADMIN'

/**
 * 화면 권한을 정하는 확인 단계입니다. 서버가 계정 상태로 계산해 내려 주며 앱은 이 값을 믿고 라우트를 지킵니다.
 * `COMPANY`는 사업자등록번호 조회를 통과한 기업을 등록하면 내려옵니다.
 */
export type AccountTier = 'MEMBER' | 'COMPANY' | 'ADMIN'

/** 환영 화면에서 고르는 회원 유형입니다. 사업자등록번호가 있으면 기업, 아직 없으면 개인입니다. */
export type AccountType = 'INDIVIDUAL' | 'BUSINESS'

/** 로그인한 계정입니다. 비밀번호·토큰은 포함하지 않습니다. */
export type Account = {
  email: string
  role: AccountRole
  tier: AccountTier
  emailVerified: boolean
  /** 등록한 기업 요약입니다. 없으면 null이며 사이드바는 이메일만 보여 줍니다. */
  company: AccountCompanySummary | null
  /** 거짓이면 소셜 로그인으로만 가입해 비밀번호가 없는 계정입니다. 프로필은 비밀번호 항목을 숨기고 계정 삭제는 비밀번호를 묻지 않습니다. */
  hasPassword: boolean
  /** 환영 화면에서 고른 회원 유형입니다. 아직이면 null입니다. */
  accountType: AccountType | null
  /** 거짓이면 로그인 뒤 다른 작업 화면보다 먼저 `/app/welcome`을 보여 줍니다. */
  onboarded: boolean
}

export type AccountCompanySummary = {
  companyName: string
  businessNumber: string
  /** 국세청 사업자 상태 코드. `01` 계속사업자만 파트너 모집글·제안을 쓸 수 있고, `02` 휴업자는 둘러보기만 됩니다. */
  businessStatusCode: BusinessStatusCode
}

/** 파트너 모집글·제안을 쓸 수 있는 기업인지입니다. 기업이 없거나 휴업이면 거짓입니다. */
export function hasActiveBusiness(account: Account): boolean {
  return account.company !== null && account.company.businessStatusCode === '01'
}

const tierRank: Record<AccountTier, number> = { MEMBER: 1, COMPANY: 2, ADMIN: 3 }

/** 계정이 요구 단계 이상인지 확인합니다. 관리자는 모든 단계를 포함합니다. */
export function meetsTier(account: Account, minimum: AccountTier): boolean {
  return tierRank[account.tier] >= tierRank[minimum]
}

File: backend/core-service/src/main/kotlin/ai/govbiz/core/account/domain/Account.kt
package ai.govbiz.core.account.domain

import java.time.LocalDateTime

/** 관리자는 SQL이나 개발용 로그인으로만 지정하며 가입 시에는 항상 USER입니다. */
enum class AccountRole {
    USER,
    ADMIN,
}

/**
 * 화면 권한을 정하는 확인 단계입니다. 역할 이름이 아니라 계정이 통과한 확인으로 계산합니다.
 *
 * `COMPANY`는 사업자등록번호 조회를 통과한 기업을 등록하면 붙습니다. 이메일 인증 조건은 인증 기능이 생길 때 더합니다.
 */
enum class AccountTier {
    MEMBER,
    COMPANY,
    ADMIN,
}

/**
 * 회원 유형입니다. 사업자등록번호가 있으면 기업, 아직 없으면 개인입니다. 환영 화면에서 한 번 고르고 프로필에서 바꿀 수 있습니다.
 * 기업 회원이라도 사업자 등록(`company`)은 별개라, 협업 기능은 등록을 마쳐야 열립니다.
 */
enum class AccountType {
    INDIVIDUAL,
    BUSINESS,
}

/** 환영 화면 2단계의 이용 목적입니다. 유형마다 고를 수 있는 값이 다르고 건너뛸 수 있습니다. */
enum class OnboardingPurpose(val allowedFor: Set<AccountType>) {
    /** 창업 지원사업 찾기 (개인) */
    FIND_STARTUP_PROGRAMS(setOf(AccountType.INDIVIDUAL)),
    /** 받을 수 있는 지원금 확인 (개인) */
    CHECK_GRANT_ELIGIBILITY(setOf(AccountType.INDIVIDUAL)),
    /** 맞는 지원사업 찾기 (기업) */
    FIND_PROGRAMS(setOf(AccountType.BUSINESS)),
    /** 함께 신청할 기업 찾기 (기업) */
    FIND_PARTNERS(setOf(AccountType.BUSINESS)),
    /** 신청 서류 준비·중복 검토 (둘 다) */
    PREPARE_DOCUMENTS(setOf(AccountType.INDIVIDUAL, AccountType.BUSINESS)),
    ;

    fun isAllowedFor(type: AccountType): Boolean = type in allowedFor
}

/** 로그인 가능한 계정입니다. 비밀번호 해시는 포함하지 않습니다. */
data class Account(
    val id: Long,
    val email: String,
    val role: AccountRole,
    val emailVerifiedAt: LocalDateTime?,
    val suspendedAt: LocalDateTime?,
    val createdAt: LocalDateTime,
    /** 등록한 기업 요약입니다. 없으면 회원(MEMBER) 단계입니다. */
    val company: CompanySummary? = null,
    /** 비밀번호를 만든 계정인지입니다. 소셜 로그인으로만 가입한 계정은 거짓이며 화면이 비밀번호 항목을 숨깁니다. */
    val hasPassword: Boolean = true,
    /** 환영 화면에서 고른 회원 유형입니다. 아직 고르지 않았으면 null입니다. */
    val accountType: AccountType? = null,
    /** 환영 화면에서 고른 이용 목적입니다. 건너뛰었거나 아직이면 null입니다. */
    val onboardingPurpose: OnboardingPurpose? = null,
    /** 환영 화면을 마친 시각입니다. null이면 로그인 뒤 `/app/welcome`을 먼저 보여 줍니다. */
    val onboardedAt: LocalDateTime? = null,
) {
    init {
        requireEmail(email)
    }

    val isOnboarded: Boolean
        get() = onboardedAt != null

    val isAdmin: Boolean
        get() = role == AccountRole.ADMIN

    val isEmailVerified: Boolean
        get() = emailVerifiedAt != null

    /** 정지된 계정은 로그인과 세션 확인이 모두 막힙니다. 정지 사유는 관리자 조치 기록이 맡습니다. */
    val isSuspended: Boolean
        get() = suspendedAt != null

    val hasCompany: Boolean
        get() = company != null

    val tier: AccountTier
        get() = when {
            isAdmin -> AccountTier.ADMIN
            hasCompany -> AccountTier.COMPANY
            else -> AccountTier.MEMBER
        }
}

/** 로그인 검증에만 쓰는 계정과 비밀번호 해시 조합입니다. 공개 계약으로 노출하지 않습니다. */
data class AccountCredential(
    val account: Account,
    val passwordHash: String,
) {
    init {
        require(passwordHash.isNotBlank()) { "passwordHash must not be blank" }
    }
}

/** DB에 저장된 로그인 세션 한 건입니다. 토큰 원문은 없고 만료·마지막 사용 시각만 있습니다. */
data class StoredAccountSession(
    val accountId: Long,
    val expiresAt: LocalDateTime,
    val lastUsedAt: LocalDateTime,
)

internal fun requireEmail(email: String) {
    require(email.isNotBlank() && email == email.trim() && email == email.lowercase()) {
        "email must be a trimmed lowercase address"
    }
}
